### PR TITLE
teuthology: add pre/post fragment merge hooks

### PR DESCRIPTION
**Pull request by batrick**
Documentation: https://github.com/batrick/teuthology/blob/yaml-preprocessor/docs/fragment_merging.rst
Ceph PR using this: https://github.com/ceph/ceph/pull/46882

Todo:
[x] Write tests
[x] Documentation

### CODE REVIEW

**Comment by batrick** [comment]
Added a Concrete Example section to the docs: https://github.com/batrick/teuthology/blob/yaml-preprocessor/docs/fragment_merging.rst

cc @jdurgin

**Comment by batrick** [comment]
With: https://github.com/ceph/ceph/pull/46882

```
$ teuthology-suite -t main --machine-type smithi --email pdonnell@redhat.com -p 75 --suite fs:verify --force-priority --filter valgrind --ceph main --ceph-repo https://github.com/ceph/ceph.git -S f9bd8922d75a454b4a0ace5a85637213e5ee5700 --suite-branch yaml-postprocess-test --suite-repo https://github.com/batrick/ceph.git
/home/pdonnell/teuthology/virtualenv/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
2022-07-14 01:58:07,754.754 INFO:teuthology.suite:Using random seed=7908
2022-07-14 01:58:07,755.755 INFO:teuthology.suite.run:kernel sha1: distro
2022-07-14 01:58:07,959.959 INFO:teuthology.suite.run:ceph sha1 explicitly supplied
2022-07-14 01:58:07,959.959 INFO:teuthology.suite.run:ceph sha1: f9bd8922d75a454b4a0ace5a85637213e5ee5700
2022-07-14 01:58:07,959.959 INFO:teuthology.suite.util:container build centos/8, checking for build_complete
2022-07-14 01:58:08,438.438 INFO:teuthology.suite.run:ceph version: 17.0.0-13521.gf9bd8922
2022-07-14 01:58:09,099.099 INFO:teuthology.suite.run:ceph branch: yaml-postprocess-test e6aebc2bae45c3856390ea5bd6fb01dc6a2ab745
2022-07-14 01:58:09,109.109 INFO:teuthology.repo_utils:Fetching yaml-postprocess-test from origin
2022-07-14 01:58:09,635.635 INFO:teuthology.repo_utils:Resetting repo at /home/pdonnell/src/github.com_batrick_ceph_yaml-postprocess-test to origin/yaml-postprocess-test
2022-07-14 01:58:10,482.482 INFO:teuthology.suite.run:teuthology branch: main 89d874d555f56b1c1f383df72487aecec0ee6cf9
2022-07-14 01:58:10,757.757 INFO:teuthology.suite.run:Suite fs:verify in /home/pdonnell/src/github.com_batrick_ceph_yaml-postprocess-test/qa/suites/fs/verify generated 240 jobs (not yet filtered or merged)
...
2022-07-14 02:02:43,765.765 INFO:teuthology.suite.run:Scheduling fs:verify/{begin/{0-install 1-ceph 2-logrotate} clusters/1a5s-mds-1c-client conf/{client mds mon osd} distro/rhel_latest_1 mount/kclient/{k-testing mount ms-die-on-skipped} objectstore-ec/bluestore-bitmap overrides/{ignorelist_health ignorelist_wrongly_marked_down mon-debug session_timeout} ranks/5 tasks/fsstress validater/valgrind/{ubuntu-overrides valgrind}}
/home/pdonnell/teuthology/virtualenv/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
Job scheduled with name pdonnell-2022-07-14_01:58:07-fs:verify-main-distro-default-smithi and ID 6929514
2022-07-14 02:02:46,751.751 INFO:teuthology.suite.run:Suite fs:verify in /home/pdonnell/src/github.com_batrick_ceph_yaml-postprocess-test/qa/suites/fs/verify scheduled 72 jobs.
2022-07-14 02:02:46,752.752 INFO:teuthology.suite.run:168/240 jobs were filtered out.
/home/pdonnell/teuthology/virtualenv/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
Job scheduled with name pdonnell-2022-07-14_01:58:07-fs:verify-main-distro-default-smithi and ID 6929515
2022-07-14 02:02:49,734.734 INFO:teuthology.suite.run:Test results viewable at http://pulpito.front.sepia.ceph.com:80/pdonnell-2022-07-14_01:58:07-fs:verify-main-distro-default-smithi/
```

https://pulpito.ceph.com/pdonnell-2022-07-14_01:58:07-fs:verify-main-distro-default-smithi/


and

```
$ teuthology-suite -t main --machine-type smithi --email pdonnell@redhat.com -p 75 --suite fs:workload  --force-priority --subset $((RANDOM % 32))/32 --ceph main --ceph-repo https://github.com/ceph/ceph.git -S f9bd8922d75a454b4a0ace5a85637213e5ee5700 --suite-branch yaml-postprocess-test --suite-repo https://github.com/batrick/ceph.git
/home/pdonnell/teuthology/virtualenv/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
2022-07-14 02:22:08,635.635 INFO:teuthology.suite:Using random seed=4192
2022-07-14 02:22:08,637.637 INFO:teuthology.suite.run:kernel sha1: distro
2022-07-14 02:22:08,852.852 INFO:teuthology.suite.run:ceph sha1 explicitly supplied
2022-07-14 02:22:08,852.852 INFO:teuthology.suite.run:ceph sha1: f9bd8922d75a454b4a0ace5a85637213e5ee5700
2022-07-14 02:22:08,852.852 INFO:teuthology.suite.util:container build centos/8, checking for build_complete
2022-07-14 02:22:09,256.256 INFO:teuthology.suite.run:ceph version: 17.0.0-13521.gf9bd8922
2022-07-14 02:22:09,808.808 INFO:teuthology.suite.run:ceph branch: yaml-postprocess-test e6aebc2bae45c3856390ea5bd6fb01dc6a2ab745
2022-07-14 02:22:09,832.832 INFO:teuthology.repo_utils:Fetching yaml-postprocess-test from origin
2022-07-14 02:22:10,249.249 INFO:teuthology.repo_utils:Resetting repo at /home/pdonnell/src/github.com_batrick_ceph_yaml-postprocess-test to origin/yaml-postprocess-test
2022-07-14 02:22:10,813.813 INFO:teuthology.suite.run:teuthology branch: main 89d874d555f56b1c1f383df72487aecec0ee6cf9
2022-07-14 02:22:10,841.841 INFO:teuthology.suite.build_matrix:Subset=14/32
2022-07-14 02:22:10,934.934 INFO:teuthology.suite.run:Suite fs:workload in /home/pdonnell/src/github.com_batrick_ceph_yaml-postprocess-test/qa/suites/fs/workload generated 140 jobs (not yet filtered or merged)
...
2022-07-14 02:26:45,909.909 INFO:teuthology.suite.run:Suite fs:workload in /home/pdonnell/src/github.com_batrick_ceph_yaml-postprocess-test/qa/suites/fs/workload scheduled 140 jobs.
2022-07-14 02:26:45,909.909 INFO:teuthology.suite.run:0/140 jobs were filtered out.
/home/pdonnell/teuthology/virtualenv/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
Job scheduled with name pdonnell-2022-07-14_02:22:08-fs:workload-main-distro-default-smithi and ID 6929657
2022-07-14 02:26:48,048.048 INFO:teuthology.suite.run:Test results viewable at http://pulpito.front.sepia.ceph.com:80/pdonnell-2022-07-14_02:22:08-fs:workload-main-distro-default-smithi/
You have new mail in /var/mail/pdonnell
```

https://pulpito.ceph.com/pdonnell-2022-07-14_02:22:08-fs:workload-main-distro-default-smithi/

cc @zmc

File: teuthology/suite/merge.py
Comment (zmc) [inline teuthology/suite/merge.py:12]: If anything ever modifies this value, we'll start to see some very mysterious issues pop up. Nothing in the PR will modify it thanks to your modifications to `deep_merge()`, but we can take an extra step to avoid future issues by making this `dict` a [`MappingProxyType`](https://docs.python.org/3/library/types.html#types.MappingProxyType). That's a trivial change, and the only other one required would be to add `or isinstance(b, MappingProxyType)` to the `assert` in `deep_merge()`.

File: teuthology/suite/test/test_merge.py
Comment (zmc) [inline teuthology/suite/test/test_merge.py:50]: Not a big deal, but `textwrap.dedent()` might be useful in cases like this

**Review by zmc — CHANGES_REQUESTED** [review]
This is a really interesting and well-though-out feature! I have a few concerns, but they should all be very easy to address.

File: teuthology/suite/run.py
Comment (batrick) [reply teuthology/suite/run.py:453]: We have a few options: put the entire yaml in a CLI argument, pass a named pipe as the arg, or maybe use stdin (not much different from a named pipe). Which do you prefer? Named pipes are clean/nice but you can only read the data once (and in order of CLI args).

File: teuthology/suite/merge.py
Comment (batrick) [reply teuthology/suite/merge.py:23]: I'll try. I would have but I wasn't feel enthusiastic about figuring out the path lookup in Python. It's probably easier than I think.

File: teuthology/suite/merge.py
Comment (batrick) [reply teuthology/suite/merge.py:12]: Makes sense!

File: teuthology/suite/test/test_merge.py
Comment (batrick) [reply teuthology/suite/test/test_merge.py:50]: I'll give it a shot, thanks.

File: teuthology/suite/merge.py
Comment (kshtsk) [reply teuthology/suite/merge.py:23]: hint: for example, Path(__file__).absolute().parent, and probably the file should be added to options.package_data setup.cfg

**Comment by batrick** [comment]
@zmc PTAL

**Review by zmc — APPROVED** [review]
Thanks for the updates! It all looks great.

I'll wait until tomorrow to merge - but if anyone sees this that's willing to monitor for unexpected issues and deal with any that arise, go for it. I don't expect any regressions.

**Pull request merged**
